docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+   }
+
+   // New
+   oe_select_attester_format(
+       requested_formats,
+       requested_formats_length,
+       &selected_format);
+```
+
+3. The application gets optional evidence parameters from verifiers and stays TEE-agnostic when generating evidences.
+
+```C
+   // Old
+   uint8_t opt_paramsX[] = { OPT_PARAMS_X };
+   uint8_t opt_paramsY[] = { OPT_PARAMS_Y };
+   if (format == ATTESTERX_UUID)
+   {
+      oe_get_evidence(&format, 0, NULL, 0, opt_paramsX, sizeof(opt_paramsX), &evidence, &evidence_size, NULL, NULL);
+   }
+   else if (format == ATTESTERY_UUID)
+   {
+       oe_get_evidence(&format, 0, NULL, 0, opt_paramsY, sizeof(opt_paramsY), &evidence, &evidence_size, NULL, NULL);
+   }
+
+   // New
+   uint8_t* opt_params;
+   size_t opt_params_size;
+   oe_get_evidence_parameters(&format, &opt_params, &opt_params_size);
+
+   oe_get_evidence(&format, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+4. SGX report (referred as local attestation in Open Enclave) and SGX quote (referred as remote attestation in Open Enclave) generation takes different inputs and generates data of different structures. The local/remote attestation concept is SGX specific. To prevent confusion at evidence generation, the OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION and OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION flags were deprecated.
+
+```C
+   // Old
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+   oe_get_evidence(&format_remote_only, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+
+   // New
+   oe_get_evidence(&format_sgx_local, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx_remote, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_remote_only, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+User Experience
+---------------
+The user experience is improved by the changes. These are a couple of possible attestation scenarios:
+
+Scenario 1 - Attester selects from a list of evidence formats that can be verified:
+- Verifier application
+    1. Upon initialization, queries available verifiers to the application (oe_sgx_get_verifiers) and registers them (oe_register_verifier).
+    2. Builds the format IDs of the registered verifiers sorted by preference.
+    3. Sends the format IDs to the attester application. 
```

Comment:
Modified scenario 2 .

---
In reply to: [393412284](https://github.com/openenclave/openenclave/pull/2621#discussion_r393412284) [](ancestors = 393412284)

Reply:
This is not about scenario 2 though, this is about whether scenario 1 is important enough that we are willing to make public API changes explicitly support. To put this in another way, I expect that there to be a sample illustrating every scenario presented in this design proposal, and I expect that many users will simply apply the samples in their code as the endorsed way to use the APIs that they illustrate.

If scenario 1 can be simply collapsed into scenario 2 by having the verifier ask the attester for evidence first, and the rest of that is exactly the same as scenario 2 afterwards, then you can basically just remove scenario 1 entirely and write a single sample for scenario 2 only. It would also suggest though that `oe_select_attester_format()` might not need to be implemented.

As a meta point, I would like to avoid a repeat of having to modify these public APIs again and I'm hoping that thinking through how you actually want to showcase the use of these APIs will prevent introducing incorrect or unnecessary APIs in retrospect. Having a clear idea of the sample implementation I think will clarify a lot of design points that may seem abstract  when simply discussing API proposals with many potential uses.